src/renderer/mod.rs

```diff
@@ -253,34 +253,106 @@ impl<'a> Renderer<'a> {
+        match *expr {
+            ExprVal::Ident(ref ident) => {
+                let v = self.lookup_ident(ident)?;
+                if v.is_i64() {
+                    Ok(Number::from(v.as_i64().unwrap()))
+                } else if v.is_u64() {
+                    Ok(Number::from(v.as_u64().unwrap()))
+                } else if v.is_f64() {
+                    Ok(Number::from_f64(v.as_f64().unwrap()).unwrap())
+                } else {
+                    bail!(
                     "Variable `{}` was used in a math operation but is not a number",
                     ident,
-                ),
+                    )
+                }
             },
-            ExprVal::Int(val) => val as f64,
-            ExprVal::Float(val) => val,
+            ExprVal::Int(val) => Ok(Number::from(val)),
+            ExprVal::Float(val) => Ok(Number::from_f64(val).unwrap()),
             ExprVal::Math(MathExpr { ref lhs, ref rhs, ref operator }) => {
                 let l = self.eval_expr_as_number(lhs)?;
                 let r = self.eval_expr_as_number(rhs)?;
+
                 match *operator {
-                    MathOperator::Mul => l * r,
-                    MathOperator::Div => l / r,
-                    MathOperator::Add => l + r,
-                    MathOperator::Sub => l - r,
-                    MathOperator::Modulo => l % r,
+                    MathOperator::Mul => {
+                        if l.is_i64() && r.is_i64() {
+                            let ll = l.as_i64().unwrap();
+                            let rr = r.as_i64().unwrap();
+                            Ok(Number::from(ll * rr))
+                        } else if l.is_u64() && r.is_u64() {
+                            let ll = l.as_u64().unwrap();
+                            let rr = r.as_u64().unwrap();
+                            Ok(Number::from(ll * rr))
+                        } else {
+                            let ll = l.as_f64().unwrap();
+                            let rr = r.as_f64().unwrap();
+			    Ok(Number::from_f64(ll * rr).unwrap())
+                        }
+                    },
+                    MathOperator::Div => {
+                        let ll = l.as_f64().unwrap();
+                        let rr = r.as_f64().unwrap();
+			let res = ll / rr;
+			if res.is_nan() {
+			    Err(From::from("NaN"))
+			} else {
+                            Ok(Number::from_f64(res).unwrap())
+			}
+                    },
+                    MathOperator::Add => { 
+                        if l.is_i64() && r.is_i64() {
+                            let ll = l.as_i64().unwrap();
+                            let rr = r.as_i64().unwrap();
+                            Ok(Number::from(ll + rr))
+                        } else if l.is_u64() && r.is_u64() {
+                            let ll = l.as_u64().unwrap();
+                            let rr = r.as_u64().unwrap();
+                            Ok(Number::from(ll + rr))
```

Comment:
I believe in all cases other than `float` we can unwrap numbers as `i64` no?

Reply:
Well, I'm pretty sure you can't cast a u64 to i64.

For example, if the input number is std::i64::MAX + 1, it is a valid u64 but definitely not the same value as an i64.

The JSON Number type from serde stores either the f64, u64 or i64 based on the value provided. We need to get the correct one. I agree it is very verbose and I hope we can improve that, I'm open to suggestion.

For reference, the implementation of Number: https://github.com/serde-rs/json/blob/master/src/number.rs